plugins/sqs_worker/worker.go

```diff
@@ -0,0 +1,154 @@
+	"go.uber.org/fx"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type AwsSqsWorker struct {
+	TopicMgr    *sqs.AwsTopicManager
+	Logger      *logger.Logger
+	HandlerMaps map[string]EventHandlerInterface
+}
+
+type awsMessage struct {
+	topicName string
+	*aws_sqs.Message
+}
+
+var enabled = true
+var started = false
+var sqsMaxMessages = int64(1)
+
+func init() {
+	plugins.Registry = append(plugins.Registry, NewAwsSqsWorker)
+}
+
+func (w *AwsSqsWorker) SetRegion(region string) {
+	w.TopicMgr.SetRegion(region)
+}
+
+func (w *AwsSqsWorker) Serve() {
+	chnMessages := make(chan *awsMessage, len(w.HandlerMaps))
+
+	for _, handlerInterface := range w.HandlerMaps {
+		topic := w.TopicMgr.GetTopic(handlerInterface.Topic().Name)
+		go w.messageLoop(chnMessages, topic)
+		w.Logger.WithFields(log.Fields{"topic": topic.Name, "queueArn": topic.Arn}).Info("Listening on SQS queue")
+	}
+
+	go func() {
+		for message := range chnMessages {
+			w.handleMessage(message)
+		}
+	}()
+
+	started = true
+}
+
+func (w *AwsSqsWorker) messageLoop(chn chan<- *awsMessage, topic *sqs.Topic) {
+	for enabled {
+		output, err := topic.ReceiveMessage(&aws_sqs.ReceiveMessageInput{
+			AttributeNames: []*string{
+				aws.String(aws_sqs.MessageSystemAttributeNameSentTimestamp),
+			},
+			MessageAttributeNames: []*string{
+				aws.String(aws_sqs.QueueAttributeNameAll),
+			},
+			QueueUrl:            &topic.Arn,
+			MaxNumberOfMessages: aws.Int64(sqsMaxMessages),
+			VisibilityTimeout:   aws.Int64(20), // 20 seconds
+			WaitTimeSeconds:     aws.Int64(0),
+		})
+
+		if err != nil {
+			w.Logger.WithFields(log.Fields{"error": err}).Error("Failed to fetch sqs message")
+		}
+
+		for _, message := range output.Messages {
+			chn <- &awsMessage{
+				topicName: topic.Name,
+				Message:   message,
+			}
+		}
+	}
+}
+
+func (w *AwsSqsWorker) handleMessage(message *awsMessage) {
+	defer func() {
+		if r := recover(); r != nil {
+			err := r.(error)
+			w.Logger.WithFields(log.Fields{"error": err}).Error("Failed to invoke event")
+
+			topic := w.TopicMgr.GetTopic(message.topicName)
+			deleteMessageInput := aws_sqs.DeleteMessageInput{QueueUrl: &topic.Arn, ReceiptHandle: message.ReceiptHandle}
+			_, _ = topic.DeleteMessage(&deleteMessageInput)
+		}
+	}()
+
+	w.Logger.WithFields(log.Fields{"message": *message.Body}).Debug("Message received")
+
+	topic := w.TopicMgr.GetTopic(message.topicName)
+	deleteMessage := true
+	err := w.HandlerMaps[topic.Name].OnEvent(topic, *message.Body)
+	if err != nil {
+		deleteMessage = w.HandlerMaps[topic.Name].OnError(topic, err)
+	}
+
+	if deleteMessage {
+		deleteMessageInput := aws_sqs.DeleteMessageInput{QueueUrl: &topic.Arn, ReceiptHandle: message.ReceiptHandle}
+		_, _ = topic.DeleteMessage(&deleteMessageInput)
+	}
+}
+
+func (w *AwsSqsWorker) Shutdown() {
+	enabled = false
+	started = false
+}
+
+func (w *AwsSqsWorker) Register(handlerInterface EventHandlerInterface) *sqs.Topic {
+	if started {
+		panic("not allow register after Serve")
```

Comment:
never panic with a string

Reply:
if possible use `github.com/pkg/errors` to create errors.
so that we will have stack trace of this line.